Optimize snapshot and report generation threads in `create_snapshots_reports_scorecard.py`

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR gives work to threads more efficiently when creating snapshots and reports in `extras/create_snapshots_reports_scorecard.py`.

Note that this does not include any changes to the way third-party snapshots and reports are generated.  That should be done in a later PR (see issue https://github.com/cisagov/cyhy-reports/issues/60).

I also took this opportunity to improve a few logging statements that did not include the thread name.  This will make the log output easier to understand.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

The current code splits the list of snapshots/reports to be generating into a group of sub-lists and assigns each thread one of those sub-lists to work on.  This PR improves upon this by enabling each thread to pull one snapshot or report (to be generated) from a single list and generating it, repeating the process until the list is empty, at which point each thread exits.  This is a more efficient way to utilize the threads.  It should result in a faster overall processing time by preventing some threads from having to generate multiple long-running snapshots/reports, while other threads are sitting idle because they have already completed their work.

This PR resolves #62.  Since I was making related changes here, I also included changes to resolve #59.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I tested this by commenting out or not executing (via switches) parts of the script that were not relevant (e.g. pausing/resuming the commander, creating the sample report, creating third-party snapshots/reports, etc).  Then I temporarily modified the script so that it ran a simple `sleep` command instead of `cyhy-snapshot` and `cyhy-report`.  This allowed me to run the entire script (e.g. `./create_snapshots_reports_scorecard.py --no-dock --no-log --no-pause cyhy-read-only scan-read-only`) and I confirmed that the threading and list consumption worked as intended.

Sample test run output (sanitized and snipped for clarity; note I also limited the number of reports here to generate to 100 for a quicker test run):
```
2024-01-26 10:29:15,075 INFO - BEGIN
2024-01-26 10:29:15,075 INFO - Building list of reports to generate...
2024-01-26 10:29:15,085 INFO - Building list of snapshots to generate...
2024-01-26 10:33:53,529 DEBUG - 100 snapshots to generate: [u'AAA', u'BBB', u'CCC', ...]
2024-01-26 10:33:53,529 DEBUG - [Thread-1] 100 snapshot(s) left to generate
2024-01-26 10:33:53,529 INFO - [Thread-1] Starting snapshot for: AAA
2024-01-26 10:33:53,562 DEBUG - [Thread-2] 99 snapshot(s) left to generate
2024-01-26 10:33:53,563 INFO - [Thread-2] Starting snapshot for: BBB
2024-01-26 10:33:53,563 DEBUG - [Thread-3] 98 snapshot(s) left to generate
2024-01-26 10:33:53,564 INFO - [Thread-3] Starting snapshot for: CCC
...
2024-01-26 10:33:54,603 INFO - [Thread-1] Successful snapshot: AAA (1.06 s)
2024-01-26 10:33:54,604 DEBUG - [Thread-1] 68 snapshot(s) left to generate
2024-01-26 10:33:54,605 INFO - [Thread-1] Starting snapshot for: YYY
2024-01-26 10:33:54,645 INFO - [Thread-3] Successful snapshot: CCC (1.04 s)
2024-01-26 10:33:54,699 INFO - [Thread-2] Successful snapshot: BBB (1.03 s)
...
2024-01-26 10:33:59,814 INFO - [Thread-1] Successful snapshot: ZZZ (2.04 s)
2024-01-26 10:33:59,814 DEBUG - [Thread-1] 0 snapshot(s) left to generate
2024-01-26 10:33:59,814 INFO - [Thread-1] No snapshots left to generate - thread exiting
...
2024-01-26 10:34:00,023 INFO - Time to complete snapshots: 4.75 minutes
2024-01-26 10:34:00,024 INFO - Longest Snapshots:
2024-01-26 10:34:00,024 INFO - GGG: 2.1 seconds
2024-01-26 10:34:00,025 INFO - HHH: 2.1 seconds
2024-01-26 10:34:00,025 INFO - MMM: 2.1 seconds
...
2024-01-26 10:34:00,026 DEBUG - 100 reports to generate: [u'AAA', u'BBB', u'CCC', ...]
2024-01-26 10:34:00,027 DEBUG - [Thread-33] 100 reports left to generate
2024-01-26 10:34:00,027 INFO - [Thread-33] Starting report for: AAA
2024-01-26 10:34:00,533 DEBUG - [Thread-34] 99 reports left to generate
2024-01-26 10:34:00,534 INFO - [Thread-34] Starting report for: BBB
2024-01-26 10:34:01,037 DEBUG - [Thread-35] 98 reports left to generate
2024-01-26 10:34:01,038 INFO - [Thread-35] Starting report for: CCC
2024-01-26 10:34:01,071 INFO - [Thread-33] Successful report generated: AAA (1.04 s)
2024-01-26 10:34:01,071 DEBUG - [Thread-33] 97 reports left to generate
2024-01-26 10:34:01,071 INFO - [Thread-33] Starting report for: FFF
...
2024-01-26 10:34:02,108 INFO - [Thread-33] Successful report generated: FFF (1.04 s)
2024-01-26 10:34:02,109 DEBUG - [Thread-33] 94 reports left to generate
2024-01-26 10:34:02,109 INFO - [Thread-33] Starting report for: JJJ
2024-01-26 10:34:02,582 INFO - [Thread-34] Successful report generated: BBB (2.05 s)
2024-01-26 10:34:02,583 DEBUG - [Thread-34] 92 reports left to generate
2024-01-26 10:34:02,583 INFO - [Thread-34] Starting report for: NNN
...
2024-01-26 10:34:03,086 INFO - [Thread-35] Successful report generated: CCC (2.05 s)
2024-01-26 10:34:03,087 DEBUG - [Thread-35] 89 reports left to generate
2024-01-26 10:34:03,087 INFO - [Thread-35] Starting report for: PPP
...
2024-01-26 10:34:14,273 INFO - [Thread-47] Successful report generated: ZZZ (2.03 s)
2024-01-26 10:34:14,273 DEBUG - [Thread-47] 0 reports left to generate
2024-01-26 10:34:14,273 INFO - [Thread-47] No reports left to generate - exiting
2024-01-26 10:34:14,274 INFO - Time to complete reports: 0.24 minutes
2024-01-26 10:34:14,274 INFO - Longest Reports:
2024-01-26 10:34:14,274 INFO - QQQ: 2.1 seconds
2024-01-26 10:34:14,274 INFO - TTT: 2.1 seconds
2024-01-26 10:34:14,274 INFO - VVV: 2.1 seconds
...
2024-01-26 10:34:14,275 INFO - Time to complete reports: 0.24 minutes
2024-01-26 10:34:14,275 INFO - Number of snapshots generated: 115
2024-01-26 10:34:14,276 INFO -   Third-party snapshots generated: 0
2024-01-26 10:34:14,276 INFO - Number of snapshots failed: 0
2024-01-26 10:34:14,276 INFO -   Third-party snapshots failed: 0
2024-01-26 10:34:14,276 INFO - Number of reports generated: 100
2024-01-26 10:34:14,276 INFO -   Third-party reports generated: 0
2024-01-26 10:34:14,277 INFO - Number of reports failed: 0
2024-01-26 10:34:14,277 INFO -   Third-party reports failed: 0
2024-01-26 10:34:14,277 INFO - Total time: 4.99 minutes
2024-01-26 10:34:14,277 INFO - END
```

I also ran a test against the entire Production set of orgs and confirmed that everything worked fine (from a thread and work consumption standpoint) with a larger set of snapshots and reports.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Deploy the updated script to Production.
- [x] Validate that the updated script runs successfully in Production.

dav3r committed:
Add global lists and thread locks for snapshots and reports

dav3r committed:
Modify function that calls cyhy-snapshot

We now pull individual org IDs from the global list of snapshots_to_generate and pass them into the create_snapshot function instead of passing in the entire list of orgs.

Each thread checks to see if the global list is empty; if it isn't, it grabs an org ID and generates a snapshot for it.  If the global list is empty, the thread exits.

dav3r committed:
Populate global snapshots_to_generate list

dav3r committed:
Call updated snapshot creation function and remove outdated code

Also, add a new debug log statement.

dav3r committed:
Add thread name to snapshot logging statements

This was overlooked in the past so I figured I'd fix it while I'm here.

dav3r committed:
Modify report generation to match snapshot generation

This means using a single global list of reports_to_generate that each thread pulls from, rather than assigning a pre-determined list of reports to each thread.

dav3r committed:
Remove a function that is no longer needed

dav3r committed:
Populate the global reports_to_generate list

jsf9k commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
             if org_descendants and not use_only_existing_snapshots:
                 logging.info(
-                    " - Includes successful descendant snapshot(s): %s", org_descendants
+                    "[%s] - Includes successful descendant snapshot(s): %s", threading.current_thread().name, org_descendants
```
I think you need an extra space here if you want to match the old, nested output:
```suggestion
                    "[%s]  - Includes successful descendant snapshot(s): %s", threading.current_thread().name, org_descendants
```

jsf9k commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
             if org_descendants and not use_only_existing_snapshots:
                 logging.error(
-                    " - Unsuccessful descendant snapshot(s): %s", org_descendants,
+                    "[%s] - Unsuccessful descendant snapshot(s): %s", threading.current_thread().name, org_descendants,
```
I think you need an extra space here if you want to match the old, nested output:
```suggestion
                    "[%s]  - Unsuccessful descendant snapshot(s): %s", threading.current_thread().name, org_descendants,
```

jsf9k commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -326,28 +337,26 @@ def generate_weekly_snapshots(db, cyhy_db_section):
     reports_to_generate = create_list_of_reports_to_generate(db)
 
     logging.info("Building list of snapshots to generate...")
+    global snapshots_to_generate
     snapshots_to_generate = create_list_of_snapshots_to_generate(
         db, reports_to_generate
     )
 
+    logging.debug("%d snapshots to generate: %s", len(snapshots_to_generate), snapshots_to_generate)
+
```
I think it's worth putting a comment here stating that we don't need to do any thread locking against `snapshots_to_generate` because the code is single-threaded at this point.  Otherwise this point could needlessly confuse us in the future.

dav3r replied:
Good thinking - please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2.

jsf9k commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -890,6 +911,7 @@ def main():
                 "No previous CybEx Scorecard JSON file found - continuing without creating CybEx Scorecard"
             )
 
+        global reports_to_generate
```
I think it's worth putting a comment here stating that we don't need to do any thread locking against `reports_to_generate` because the code is single-threaded at this point.  Otherwise this point could needlessly confuse us in the future.

dav3r replied:
Good thinking - please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2.

jsf9k reviewed (changes requested):
Requesting a few minor changes.

It's probably also worth at least creating an issue for more gracefully handling snapshot or report generation failures.  Right now I think the corresponding thread will crash and exit its thread loop and we will lose that bit of concurrency.  At a minimum we should do some error handling to ensure that the thread does not exit, and eventually we probably want to put the item back onto the queue for another chance before abandoning all hope.

dav3r committed:
Add a space to preserve nested output formatting

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Add comments explaining why we don't need thread locks in some spots

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r commented:
> It's probably also worth at least creating an issue for more gracefully handling snapshot or report generation failures. Right now I think the corresponding thread will crash and exit its thread loop and we will lose that bit of concurrency. At a minimum we should do some error handling to ensure that the thread does not exit, and eventually we probably want to put the item back onto the queue for another chance before abandoning all hope.

I don't believe we will lose the thread if a failure occurs during snapshot or report generation because those commands (`cyhy-snapshot` and `cyhy-report`) are [run in a sub-process](https://github.com/cisagov/cyhy-reports/blob/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2/extras/create_snapshots_reports_scorecard.py#L264-L269), whose [return code is checked by the thread](https://github.com/cisagov/cyhy-reports/blob/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2/extras/create_snapshots_reports_scorecard.py#L290-L309).  If I'm misunderstanding you or you think I'm wrong, please let me know.

I think that if a failure occurs, we will end up in the same position as we are now (before this PR).  Namely, we will see in the output that there was one or more failed snapshots/reports and we will have to manually attempt to generate them later.  I will still create an issue to document this where we can discuss options for automatically re-attempting the failed snapshots/reports before giving up.

jsf9k commented:
> > It's probably also worth at least creating an issue for more gracefully handling snapshot or report generation failures. Right now I think the corresponding thread will crash and exit its thread loop and we will lose that bit of concurrency. At a minimum we should do some error handling to ensure that the thread does not exit, and eventually we probably want to put the item back onto the queue for another chance before abandoning all hope.
> 
> I don't believe we will lose the thread if a failure occurs during snapshot or report generation because those commands (`cyhy-snapshot` and `cyhy-report`) are [run in a sub-process](https://github.com/cisagov/cyhy-reports/blob/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2/extras/create_snapshots_reports_scorecard.py#L264-L269), whose [return code is checked by the thread](https://github.com/cisagov/cyhy-reports/blob/c2a99d6b63fbba6c4b4a6072600c02e2c6b85fd2/extras/create_snapshots_reports_scorecard.py#L290-L309). If I'm misunderstanding you or you think I'm wrong, please let me know.
> 
> I think that if a failure occurs, we will end up in the same position as we are now (before this PR). Namely, we will see in the output that there was one or more failed snapshots/reports and we will have to manually attempt to generate them later. I will still create an issue to document this where we can discuss options for automatically re-attempting the failed snapshots/reports before giving up.

Nope, I think you are correct.  _IGNORE ME!!!_

jsf9k reviewed (approved):
Nice work!

dav3r commented:
> I will still create an issue to document this where we can discuss options for automatically re-attempting the failed snapshots/reports before giving up.

See https://github.com/cisagov/cyhy-reports/issues/107 for this.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
 
 
-def create_snapshots_from_list(org_list, db, cyhy_db_section):
+def create_all_snapshots(db, cyhy_db_section):
     """Create a snapshot for each organization in a list."""
```
Not a big fan of the new function name but more importantly I would like the function docstring to better reflect what the function is now doing. I think the name is unappealing because it doesn't really reflect the threaded use of this function. Since we're no longer passing in a list of orgs I think the docstring should make some reference to the global list being consumed.

dav3r replied:
That's fair.  Please have a look at https://github.com/cisagov/cyhy-reports/pull/106/commits/e969421ea40668481a76a06786db823b27d54eb9 and let me know what you think.  Once we come to consensus on this, I will make similar changes for the report generation functions (which will also resolve #59).

mcdonnnj replied:
Ah yes, that reads much better. Thank you for being comprehensive about your approach!

jsf9k replied:
@dav3r - I have one minor quibble.  Please see [here](https://github.com/cisagov/cyhy-reports/pull/106#pullrequestreview-1854813077).

dav3r replied:
Done and done. 👍

dav3r replied:
As mentioned earlier in this conversation, please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c410e90e6457f84bfefe22006790ca0e7a91d910 where I cleaned up the report generation functions along the same lines as what I did in https://github.com/cisagov/cyhy-reports/commit/e969421ea40668481a76a06786db823b27d54eb9.

@mcdonnnj I think we can now resolve this conversation if you're satisfied.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
-    for org_id in org_list:
+    while True:
+        with stg_lock:
+            global snapshots_to_generate
```
Does this need to happen for every iteration of the loop?

dav3r replied:
You're right- please see https://github.com/cisagov/cyhy-reports/pull/106/commits/2143a3e33e22096fcca7766af83d7792166e46db.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
+    while True:
+        with stg_lock:
+            global snapshots_to_generate
+            logging.debug(
+                "[%s] %d snapshot(s) left to generate", threading.current_thread().name, len(snapshots_to_generate))
```
```suggestion
            logging.debug(
                "[%s] %d snapshot(s) left to generate",
                threading.current_thread().name,
                len(snapshots_to_generate),
            )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
+            if snapshots_to_generate:
+                org_id = snapshots_to_generate.pop(0)
+            else:
+                logging.info(
+                "[%s] No snapshots left to generate - thread exiting", threading.current_thread().name)
```
```suggestion
                logging.info(
                    "[%s] No snapshots left to generate - thread exiting",
                    threading.current_thread().name,
                )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
             if org_descendants and not use_only_existing_snapshots:
                 logging.info(
-                    " - Includes successful descendant snapshot(s): %s", org_descendants
+                    "[%s]  - Includes successful descendant snapshot(s): %s", threading.current_thread().name, org_descendants
```
```suggestion
                    "[%s]  - Includes successful descendant snapshot(s): %s",
                    threading.current_thread().name,
                    org_descendants,
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
 
     if snapshot_process.returncode == 0:
-        logging.info("Successful snapshot: %s (%.2f s)", org_id, snapshot_duration)
+        logging.info("[%s] Successful snapshot: %s (%.2f s)", threading.current_thread().name, org_id, snapshot_duration)
```
```suggestion
        logging.info(
            "[%s] Successful snapshot: %s (%.2f s)",
            threading.current_thread().name,
            org_id,
            snapshot_duration,
        )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
                 successful_snapshots.extend(org_descendants)
     else:
-        logging.error("Unsuccessful snapshot: %s", org_id)
+        logging.error("[%s] Unsuccessful snapshot: %s", threading.current_thread().name, org_id)
```
```suggestion
        logging.error(
            "[%s] Unsuccessful snapshot: %s", threading.current_thread().name, org_id
        )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
             if org_descendants and not use_only_existing_snapshots:
                 logging.error(
-                    " - Unsuccessful descendant snapshot(s): %s", org_descendants,
+                    "[%s]  - Unsuccessful descendant snapshot(s): %s", threading.current_thread().name, org_descendants,
```
```suggestion
                    "[%s]  - Unsuccessful descendant snapshot(s): %s",
                    threading.current_thread().name,
                    org_descendants,
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,30 +288,41 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
                 )
                 failed_snapshots.extend(org_descendants)
-        logging.error("Stderr failure detail: %s %s", data, err)
+        logging.error("[%s] Stderr failure detail: %s %s", threading.current_thread().name, data, err)
```
```suggestion
        logging.error(
            "[%s] Stderr failure detail: %s %s",
            threading.current_thread().name,
            data,
            err,
        )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -326,28 +337,28 @@ def generate_weekly_snapshots(db, cyhy_db_section):
         db, reports_to_generate
     )
 
+    logging.debug("%d snapshots to generate: %s", len(snapshots_to_generate), snapshots_to_generate)
```
```suggestion
    logging.debug(
        "%d snapshots to generate: %s",
        len(snapshots_to_generate),
        snapshots_to_generate,
    )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -495,27 +514,31 @@ def gen_weekly_reports(
-                target=create_reports,
-                args=(i, cyhy_db_section, scan_db_section, use_docker, nolog),
+            report_thread = threading.Thread(
+                target=create_all_reports, args=(cyhy_db_section, scan_db_section, use_docker, nolog)
```
```suggestion
                target=create_all_reports,
                args=(cyhy_db_section, scan_db_section, use_docker, nolog),
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+def create_report(org_id, cyhy_db_section, scan_db_section, use_docker, nolog):
+    """Create a report for a single organization."""
+    report_time = time.time()
+    logging.info("[%s] Starting report for: %s", threading.current_thread().name, org_id)
```
```suggestion
    logging.info(
        "[%s] Starting report for: %s", threading.current_thread().name, org_id
    )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+    while True:
+        with rtg_lock:
+            global reports_to_generate
+            logging.debug(
+                "[%s] %d reports left to generate", threading.current_thread().name, len(reports_to_generate))
```
```suggestion
            logging.debug(
                "[%s] %d reports left to generate",
                threading.current_thread().name,
                len(reports_to_generate),
            )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+            if reports_to_generate:
+                org_id = reports_to_generate.pop(0)
+            else:
+                logging.info(
+                "[%s] No reports left to generate - exiting", threading.current_thread().name)
```
```suggestion
                logging.info(
                    "[%s] No reports left to generate - exiting",
                    threading.current_thread().name,
                )
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+        failed_reports.append(org_id)
+
+
+def create_all_reports(cyhy_db_section, scan_db_section, use_docker, nolog):
+    """Create reports for all orgs in a list."""
```
Same comment about the name/docstring as for `create_all_snapshots()`.

dav3r replied:
Please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c410e90e6457f84bfefe22006790ca0e7a91d910 where I cleaned up the report generation functions along the same lines as what I did for snapshots in https://github.com/cisagov/cyhy-reports/commit/e969421ea40668481a76a06786db823b27d54eb9.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+    """Create reports for all orgs in a list."""
+    while True:
+        with rtg_lock:
+            global reports_to_generate
```
Same question about every iteration.

dav3r replied:
You're right- please see https://github.com/cisagov/cyhy-reports/pull/106/commits/2143a3e33e22096fcca7766af83d7792166e46db.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+            data,
+            err,
+        )
+        failed_reports.append(org_id)
```
I know this is inherited but is there a reason we're not using `fr_lock` before trying to modify the list?

dav3r replied:
Excellent catch - please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c13a86bb3965934b98f7e5f114c6d1c0616899d5.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+            org_id,
+            round(report_time, 2),
+        )
+        successful_reports.append(org_id)
```
I know this is inherited but is there a reason we're not using `sr_lock` before trying to modify the list?

dav3r replied:
Excellent catch - please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c13a86bb3965934b98f7e5f114c6d1c0616899d5.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -365,127 +376,135 @@ def generate_weekly_snapshots(db, cyhy_db_section):
+            )
+    data, err = p.communicate()
+    report_time = time.time() - report_time
+    report_durations.append((org_id, report_time))
```
Again but this time about `rd_lock`.

dav3r replied:
Excellent catch - please see https://github.com/cisagov/cyhy-reports/pull/106/commits/c13a86bb3965934b98f7e5f114c6d1c0616899d5.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -890,6 +913,9 @@ def main():
                 "No previous CybEx Scorecard JSON file found - continuing without creating CybEx Scorecard"
             )
 
+        global reports_to_generate
```
Loop iterations question.

dav3r replied:
This one is not in a loop, so I think it's a false alarm.

mcdonnnj replied:
Ah you are correct. My mistake!

mcdonnnj reviewed (commented):
I'm going to do one last pass on the logic but I wanted to throw out this feedback instead of sitting on it until I finish my review. Mostly around manual `black`ening and formatting with some questions/requests interspersed.

dav3r committed:
Clean up code formatting

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Move global list declarations outside of loop

This only needs to be done once before the loop starts, not on every iteration of the loop.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Update function names, docstrings, and comments to be more meaningful

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add missing thread locks to protect global list modifications

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove import statement for a library that is no longer used

jsf9k commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,66 +287,105 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
-    """Create a snapshot for each organization in a list."""
-    for org_id in org_list:
+def generate_snapshots_from_list(db, cyhy_db_section):
+    """Attempt to generate a snapshot for each organization in a global list.
+    Each thread pulls an organization ID from the global list
+    (snapshots_to_generate) and attempts to generate a snapshot for it."""
```
I think pydoc string should really be like this:
```suggestion
    """Attempt to generate a snapshot for each organization in a global list.

    Each thread pulls an organization ID from the global list
    (snapshots_to_generate) and attempts to generate a snapshot for it."""
```

jsf9k reviewed (changes requested):
One minor formatting thang.

dav3r committed:
Use a more standard format for a pydoc string

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Update report function names, docstrings, and comments to be more meaningful

This also makes them named and structured similarly to the snapshot functions.

While I was here, I tidied up the code that formulates the report generation command since it was a bit bulky before.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -288,66 +287,106 @@ def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
-def generate_weekly_snapshots(db, cyhy_db_section):
-    """Generate all snapshots needed in order to generate the CyHy reports."""
+def manage_snapshot_threads(db, cyhy_db_section):
+    """Build the lists of reports and snapshots to be generated, then spawn the
+    threads that generate the snapshots."""
```
In the same vein as https://github.com/cisagov/cyhy-reports/pull/106#discussion_r1473450650 would you reformat this docstring please.

dav3r replied:
@mcdonnnj Can you please clarify how you would like this one to be reformatted?  The comment you linked to had two sentences and @jsf9k asked me to split them up with a blank line in between.  This pydoc description is a single sentence, so I'm not clear on what you'd like me to do here.  Are you asking me to remove the line break so that the pydoc string is on a single line?  Or something else?

dav3r replied:
After chatting with @jsf9k, we came up with https://github.com/cisagov/cyhy-reports/pull/106/commits/aceb30388e15fd0171e0d6f907b327d66affdfde.  Please resolve this convo if that works for you.

mcdonnnj replied:
Apologies for being unclear with my feedback. That looks perfect and in-line with the intent behind my comment.

jsf9k replied:
@mcdonnnj and I are on the same wavelength.  :laughing:

mcdonnnj reviewed (approved):
Thanks for clearing up my feedback. This LGTM with one minor quibble around a docstring.

dav3r committed:
Improve pydoc string formatting

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
